contracts/utils/TimestampedHashRegistry.sol

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "./interfaces/ITimestampedHashRegistry.sol";
+
+// TODO: would SelfMulticall be useful?
+contract TimestampedHashRegistry is Ownable, EIP712, ITimestampedHashRegistry {
+    using ECDSA for bytes32;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    mapping(bytes32 => EnumerableSet.AddressSet) private _hashTypeToSigners;
+
+    mapping(bytes32 => SignedHash) public hashTypeToSignedHash;
+
+    bytes32 private constant _SIGNED_HASH_TYPE_HASH =
+        keccak256(
+            "SignedHash(bytes32 hashType,bytes32 hash,uint256 timestamp)"
+        );
+
+    constructor() EIP712("TimestampedHashRegistry", "1.0.0") {}
+
+    function setHashTypeSigners(
+        bytes32 hashType,
+        address[] calldata signers
+    ) external onlyOwner {
+        require(hashType != bytes32(0), "Hash is zero");
+        require(signers.length != 0, "Signers length is empty");
+        for (uint256 ind = 0; ind < signers.length; ind++) {
+            _hashTypeToSigners[hashType].add(signers[ind]);
+        }
+        emit SetHashTypeSigners(hashType, signers);
+    }
+
+    function registerSignedHash(
+        bytes32 hashType,
+        SignedHash calldata signedHash,
+        bytes[] calldata signatures
+    ) external {
+        require(hashType != bytes32(0), "Hash type is zero");
+        EnumerableSet.AddressSet storage signers = _hashTypeToSigners[hashType];
+        require(signers.length() != 0, "Signers have not been set");
+        require(
+            signatures.length == signers.length(),
+            "Signatures length mismatch"
+        );
+        for (uint256 ind = 0; ind < signatures.length; ind++) {
+            require(
+                signers.contains(
```

Comment:
This is kinda expensive. The DapiFallbackV1 version which expected the signatures to be ordered correctly was a good tradeoff imo.

Reply:
> This is kinda expensive. The DapiFallbackV1 version which expected the signatures to be ordered correctly was a good tradeoff imo.

I was really proud with achieving not having to rely on the caller sending the signatures in an expected order 😅 

> btw there's nothing preventing the caller from using the same signature multiple times

I'm confused by this. Wouldn't the signature check fail if a user sends a signature for a different `root` or `timestamp`? Or are you suggesting that I check the `root` parameter is different from a previously registered root for a hash type?

Reply:
I mean if I have one valid signature, I can repeat it `signers.length` times to create a `signatures` array that will pass this verification. Note that a signer create multiple valid signatures of the same hash (signature malleability), which means you can't avoid this by preventing signatures from being repeated. All in all, you will have to rely on the caller sending the signatures in the expected order (or sort it for them in the contract), which means the EnumerableSet is only good for having `addSigner()`/`removeSigner()` functions.

Reply:
Got it. Good catch 👏🏻 

An alternative that I can think of is creating a new `Enumerable.AddressSet` for all addresses derived from the `signatures`. Then loop over the `signers` set and use `contains` on the derived addresses set to check if a signature for the signer is present. This would allow the caller to send the signatures in any order but it might be more expensive.

Reply:
[This](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol#L134-L135) is making me uneasy tho 🤔

Reply:
I think using any design other than Safe's is unnecessarily adventurous here. Keep the owners in an `AddressSet`, have [`setupSigners()`](https://github.com/safe-global/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L28), [`addSigner()`](https://github.com/safe-global/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L58), [`removeSigner()`](https://github.com/safe-global/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L78) functions. [Require signatures to be correctly ordered](https://github.com/safe-global/safe-contracts/blob/main/contracts/Safe.sol#L295). The CI will ensure that [our metadata](https://github.com/api3dao/manager-multisig/blob/main/data/arbitrum-goerli-testnet/metadata.json#L2) matches the on-chain order anyway and we can use that to order the signatures accordingly (btw currently the manager-multisig CI doesn't ensure this for [Safe](https://github.com/api3dao/manager-multisig/blob/main/scripts/src/verify-metadata.ts#L34) or [DapiFallbackV1](https://github.com/api3dao/manager-multisig/blob/main/scripts/src/verify-metadata.ts#L66), this needs to be fixed).

Reply:
> have [`setupSigners()`](https://github.com/safe-global/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L28), [`addSigner()`](https://github.com/safe-global/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L58), [`removeSigner()`](https://github.com/safe-global/safe-contracts/blob/main/contracts/base/OwnerManager.sol#L78) functions

`setupSigners()` should completely override the elements in the set, right? This is an external call that should probably be called after deploying the contract to initialize the set of signers.

> [Require signatures to be correctly ordered](https://github.com/safe-global/safe-contracts/blob/main/contracts/Safe.sol#L295). 

`removeSigner()` would change the order of the set since [EnumerableSet._remove()](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol#L89-L91) swaps the last element of the set with the one being removed 😕 (btw [signatureSplit()](https://github.com/safe-global/safe-contracts/blob/main/contracts/Safe.sol#L295) only extracts the r,s,v parts from a signature at a specified position but this does not ensure that the actual signature order against the previously set owners list, I probably didn't understand the reference tho 🤭 )

> The CI will ensure that [our metadata](https://github.com/api3dao/manager-multisig/blob/main/data/arbitrum-goerli-testnet/metadata.json#L2) matches the on-chain order anyway and we can use that to order the signatures accordingly 

Does this mean that after calling `removeSigner()` and the order of the signers changes on-chain we should expect an error in CI and then fix the signers order in our metadata?

Reply:
> owners == signers, right?

Right

> setupSigners() should completely override the elements in the set, like a reset? or should this be an external function that should only be called once for each hashType to initialize the set of signers (if signers have ever been set for the hashType then revert similar to Safe.sol [setup()](https://github.com/safe-global/safe-contracts/blob/main/contracts/Safe.sol#L106)).

The latter

> removeSigner() would change the order of the set since [EnumerableSet._remove()](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol#L89-L91) swaps the last element of the set with the one being removed 😕

Not a problem. It's the users' (or maybe the manager-multisig frontend's) responsibility to update the metadata accordingly.

> (btw [signatureSplit()](https://github.com/safe-global/safe-contracts/blob/main/contracts/Safe.sol#L295) only extracts the r,s,v parts from a signature at a specified position but this does not ensure the actual signature order against the previously set owners list, I probably didn't understand the reference tho 🤭 )

I mean Safe also depends on sending the signatures in the correct order, sending them in an array or in serialized form doesn't matter

> Does this mean that after calling removeSigner() and the order of the signers changes on-chain we should expect an error in CI and then fix the signers order in our metadata?

Either that or the frontend should have a "create tx to remove signer" button that you click that updates the metadata as expected and create the transaction at the same time. Once the transaction is executed the CI passes automatically.